test/e2e/providers/vsphere/vsphere.go

```diff
@@ -0,0 +1,140 @@
+package vsphere
+
+import (
+	"encoding/json"
+	"fmt"
+
+	mapi "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	vsphere "github.com/openshift/machine-api-operator/pkg/apis/vsphereprovider/v1beta1"
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/openshift/windows-machine-config-operator/test/e2e/clusterinfo"
+)
+
+const defaultCredentialsSecretName = "vsphere-cloud-credentials"
+
+// Provider is a provider struct for testing vSphere
+type Provider struct {
+	oc *clusterinfo.OpenShift
+}
+
+// New returns a new vSphere provider struct with the given client set and ssh key pair
+func New(clientset *clusterinfo.OpenShift) (*Provider, error) {
+	return &Provider{
+		oc: clientset,
+	}, nil
+}
+
+// newVSphereMachineProviderSpec returns a vSphereMachineProviderSpec generated from the inputs, or an error
+func newVSphereMachineProviderSpec(clusterID string) (*vsphere.VSphereMachineProviderSpec, error) {
+	if clusterID == "" {
+		return nil, fmt.Errorf("clusterID is empty")
+	}
+
+	return &vsphere.VSphereMachineProviderSpec{
+		TypeMeta: meta.TypeMeta{
+			APIVersion: "vsphereprovider.openshift.io/v1beta1",
+			Kind:       "VSphereMachineProviderSpec",
+		},
+		CredentialsSecret: &core.LocalObjectReference{
+			Name: defaultCredentialsSecretName,
+		},
+		DiskGiB:   int32(128),
+		MemoryMiB: int64(16384),
+		Network: vsphere.NetworkSpec{
+			Devices: []vsphere.NetworkDeviceSpec{{NetworkName: "ci-segment"}},
+		},
+		NumCPUs:           int32(4),
+		NumCoresPerSocket: int32(1),
+		// The template is hardcoded with an image which has been properly sysprepped.
+		// TODO: Find a way to automatically update this with latest image
+		Template: "1909-template-with-docker-ssh",
+		Workspace: &vsphere.Workspace{
+			Datacenter:   "SDDC-Datacenter",
+			Datastore:    "WorkloadDatastore",
+			Folder:       "/SDDC-Datacenter/vm/" + clusterID,
+			ResourcePool: "/SDDC-Datacenter/host/Cluster-1/Resources",
+			Server:       "vcenter.sddc-44-236-21-251.vmwarevmc.com",
+		},
+	}, nil
+}
+
+// GenerateMachineSet generates the machineset object which is aws provider specific
+func (p *Provider) GenerateMachineSet(withWindowsLabel bool, replicas int32) (*mapi.MachineSet, error) {
+	clusterID, err := p.oc.GetInfrastructureID()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get cluster id: %v", err)
+	}
+
+	// create new machine provider spec for deploying Windows node in the same Location and Zone as master-0
+	providerSpec, err := newVSphereMachineProviderSpec(clusterID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new azure machine provider spec: %v", err)
```

Comment:
`failed to create new vsphere machine provider spec `

Reply:
`Machine`

Reply:
Please use `errors.Wrapf`. Applies to all places where you are returning `fmt.Errorf`.